### PR TITLE
Raise `ClientDisconnected` on `send()` when client disconnected

### PR DESCRIPTION
**Pull request by Kludex**
~We need to clarify this on https://github.com/django/asgiref/commit/3f0147dc79bae052113b31a02db4f2bdfffc7d4e before merging it.~

👆 Clarified.

Also, tests are missing here.

### CODE REVIEW

File: uvicorn/protocols/http/httptools_impl.py [hunk 1]
```diff
@@ -414,11 +414,13 @@ def __init__(
...
 
     # ASGI exception wrapper
-    async def run_asgi(self, app: "ASGI3Application") -> None:
+    async def run_asgi(self, app: ASGI3Application) -> None:
         try:
             result = await app(  # type: ignore[func-returns-value]
                 self.scope, self.receive, self.send
             )
+        except ClientDisconnected:
+            ...
```
Comment (Kludex) [inline uvicorn/protocols/http/httptools_impl.py]: Do I need to call the `self.transport.close()`?

File: uvicorn/protocols/http/h11_impl.py [hunk 2]
```diff
@@ -412,6 +413,8 @@ async def run_asgi(self, app: "ASGI3Application") -> None:
             result = await app(  # type: ignore[func-returns-value]
                 self.scope, self.receive, self.send
             )
+        except ClientDisconnected:
+            ...
```
Comment (Kludex) [inline uvicorn/protocols/http/h11_impl.py:417]: Do I need to call `self.transport.close()`?

File: uvicorn/protocols/http/httptools_impl.py (hunk 1 above)
Comment (Kludex) [reply uvicorn/protocols/http/httptools_impl.py]: Well... The `connection_lost` callback is called, and the transport is closed there.

Comment (Kludex) [reply uvicorn/protocols/http/httptools_impl.py]: Actually, not adding the `self.transport.close()` means that we maintain the same behavior as before (which is better).

**Comment by Kludex** [comment]
I'm going to add a test before merging, but review welcome.

File: uvicorn/protocols/http/h11_impl.py (hunk 2 above)
Comment (adriangb) [inline uvicorn/protocols/http/h11_impl.py]:
```suggestion
            pass
```

I think is more common in this context?

**Review by adriangb — COMMENTED** [review]
Looks good, just missing a test

File: tests/protocols/test_http.py [hunk 3]
```diff
@@ -629,9 +620,26 @@ async def app(scope: Scope, receive: ASGIReceiveCallable, send: ASGISendCallable
 
 
 @pytest.mark.anyio
-async def test_early_response(
-    http_protocol_cls: HTTPProtocol,
-):
+async def test_disconnect_on_send(http_protocol_cls: HTTPProtocol) -> None:
```
Comment (Kludex) [inline tests/protocols/test_http.py:603]: Test added.

**Review by adriangb — APPROVED** [review]

**Pull request merged**